x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/TraceSummary.tsx

```diff
@@ -0,0 +1,255 @@
+  font-weight: bold;
+`;
+
+interface HttpStatusBadgeProps {
+  status?: number;
+}
+const HttpStatusBadge: React.FunctionComponent<HttpStatusBadgeProps> = ({
+  status
+}) => {
+  let color = 'default';
+
+  if (!status) {
+    return null;
+  } else if (status < 300) {
+    color = theme.euiColorVis0;
+  } else if (status >= 300 || status < 400) {
+    color = theme.euiColorVis5;
+  } else if (status >= 400 || status < 500) {
+    color = theme.euiColorVis7;
+  } else if (status >= 500) {
+    color = theme.euiColorVis2;
+  }
+
+  return (
+    <EuiToolTip
+      content={i18n.translate('xpack.apm.transactionDetails.statusCode', {
+        defaultMessage: 'Status code'
+      })}
+    >
+      <EuiBadge color={color}>{status}</EuiBadge>
+    </EuiToolTip>
+  );
+};
+
+const HttpInfoBadge = styled(EuiBadge)`
+  max-width: 250px;
+`;
+
+interface HttpInfoProps {
+  transaction: Transaction;
+}
+const HttpInfo: React.FunctionComponent<HttpInfoProps> = ({ transaction }) => {
+  const method = (
+    idx(transaction, _ => _.http.request.method) || ''
+  ).toUpperCase();
+  const isRumAgent = isRumAgentName(transaction.agent.name);
+  const url = isRumAgent
+    ? idx(transaction, _ => _.transaction.page.url) || NOT_AVAILABLE_LABEL
+    : idx(transaction, _ => _.url.full) || NOT_AVAILABLE_LABEL;
+  const status = idx(transaction, _ => _.http.response.status_code);
+
+  return (
+    <>
+      <HttpInfoBadge title={undefined}>
+        <EuiToolTip
+          content={i18n.translate(
+            'xpack.apm.transactionDetails.requestMethodLabel',
+            {
+              defaultMessage: 'Request method'
+            }
+          )}
+        >
+          <MethodText>{method}</MethodText>
+        </EuiToolTip>{' '}
+        <EuiToolTip content={url}>
+          <>{url}</>
+        </EuiToolTip>
+      </HttpInfoBadge>{' '}
+      <HttpStatusBadge status={status} />
+    </>
+  );
+};
+
+interface ResultProps {
+  result: string;
+}
+const Result: React.FunctionComponent<ResultProps> = ({ result }) => {
+  return (
+    <EuiToolTip
+      content={i18n.translate('xpack.apm.transactionDetails.resultLabel', {
+        defaultMessage: 'Result'
+      })}
+    >
+      <EuiBadge color="default" title={undefined}>
+        {result}
+      </EuiBadge>
+    </EuiToolTip>
+  );
+};
+
+interface Props {
+  errorCount: number;
+  totalDuration?: number;
+  transaction: Transaction;
+}
+
+export const TraceSummary: React.FunctionComponent<Props> = ({
+  errorCount,
+  totalDuration,
+  transaction
+}) => {
+  const isHttp = 'http' in transaction;
+  const result = idx(transaction, _ => _.transaction.result);
+
+  return (
+    <>
+      <Heading />
+      <EuiText size="s">
+        <Timestamp transactionTimestamp={transaction['@timestamp']} />
+        <Separator />
+        <Duration
+          duration={transaction.transaction.duration.us}
+          totalDuration={totalDuration}
+        />
+        {(isHttp || result) && <Separator />}
```

Comment:
I think this can be improved a little. The conditions and different branches make it a little hard to follow for me. What do you think of:

```
const items = [
    <Timestamp transactionTimestamp={transaction['@timestamp']} />,
    <Duration
      duration={transaction.transaction.duration.us}
      totalDuration={totalDuration}
    />
  ];

  if (isHttp) {
    items.push(<HttpInfo transaction={transaction} />);
  } else if (result) {
    items.push(<Result result={result} />);
  }

  if (errorCount > 0) {
    items.push(
      <ErrorCountBadge
        errorCount={errorCount}
        includeText={true}
        title={undefined}
      />
    );
  }
```

and then:

```ts
{items.reduce(
  (acc, item, index) => {
    const next = acc.concat(item);
    if (index !== items.length - 1) {
      next.push(<Separator />);
    }
    return next;
  },
  [] as React.ReactElement[]
)}
```

I think this would improve further if you use a EUIFlexItem with padding/border-right too.

Reply:
I find that much harder to follow, but I'll try to make what I have there easier to follow.